### PR TITLE
AMQ-8472 - Switch to reload4j for logging

### PR DESCRIPTION
**Pull request by coheigea**

### CODE REVIEW

**Comment by jbonofre** [comment]
While Jenkins is running a build, I'm doing a local build and tests.

**Comment by jbonofre** [comment]
@coheigea we will have probably to do a force rebase as I cancel the vote (and change the git history).

**Comment by coheigea** [comment]
Sure, whatever is needed, let me know if you need anything from me on it.

**Comment by gemmellr** [comment]
Running maven dependency:tree with this change in place still shows lots of instances of log4j and slf4j-log4j12 still present in the build. Some are provided scope and so wont be passed on, but not all, and either way they still shouldn't exist. Probably some more exclusions and/or dependencyManagement needed.

**Comment by coheigea** [comment]
@gemmellr I don't see any of these when I run the dependency tree on this branch:

➜  activemq git:(AMQ-8472) mvn dependency:tree | grep slf4j-log4j
➜  activemq git:(AMQ-8472) mvn dependency:tree | grep log4j
[INFO] ------------< org.apache.activemq:activemq-log4j-appender >-------------
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ activemq-log4j-appender ---
[INFO] org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:compile
[INFO] |  +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:test
[INFO] |  |  +- org.ops4j.pax.logging:pax-logging-log4j2:jar:1.11.9:test
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:compile
[INFO] +- io.fabric8.insight:insight-log4j:jar:1.2.0.Beta4:compile

**Comment by gemmellr** [comment]
I see a bunch:
```
$ mvn dependency:tree | grep log4j
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:provided
[INFO] |  +- log4j:log4j:jar:1.2.17:provided
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:provided
[INFO] |  +- log4j:log4j:jar:1.2.17:provided
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |  +- log4j:log4j:jar:1.2.17:compile
[INFO] |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |     +- log4j:log4j:jar:1.2.17:compile
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] ------------< org.apache.activemq:activemq-log4j-appender >-------------
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ activemq-log4j-appender ---
[INFO] org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:compile
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |     +- log4j:log4j:jar:1.2.17:compile
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:test
[INFO] |  |  +- org.ops4j.pax.logging:pax-logging-log4j2:jar:1.11.9:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |  +- log4j:log4j:jar:1.2.17:compile
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:compile
[INFO] +- io.fabric8.insight:insight-log4j:jar:1.2.0.Beta4:compile
```

E.g the first is in the activemq-leveldb-store module:
```
<snipped>
[INFO] +- org.apache.zookeeper:zookeeper:jar:3.4.14:provided
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:provided
[INFO] |  +- com.github.spotbugs:spotbugs-annotations:jar:3.1.9:provided
[INFO] |  +- log4j:log4j:jar:1.2.17:provided
[INFO] |  +- jline:jline:jar:0.9.94:provided
[INFO] |  +- org.apache.yetus:audience-annotations:jar:0.5.0:provided
[INFO] |  \- io.netty:netty:jar:3.10.6.Final:provided
```

Its visible from the `mvn help:effective-pom` output for that module that it does have a dependency on zookeper 3.4.14 and does not exclude log4j or slf4j-log4j12 from it, and so is pulling them in since zookeper does depend on both:
https://repo1.maven.org/maven2/org/apache/zookeeper/zookeeper/3.4.14/zookeeper-3.4.14.pom

**Comment by jbonofre** [comment]
Here the full output:

```
[INFO] ActiveMQ :: Log4j Appender                                         [jar]
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:provided
[INFO] |  +- log4j:log4j:jar:1.2.17:provided
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:provided
[INFO] |  +- log4j:log4j:jar:1.2.17:provided
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |  +- log4j:log4j:jar:1.2.17:compile
[INFO] |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |     +- log4j:log4j:jar:1.2.17:compile
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] ------------< org.apache.activemq:activemq-log4j-appender >-------------
[INFO] Building ActiveMQ :: Log4j Appender 5.16.4-SNAPSHOT              [26/42]
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ activemq-log4j-appender ---
[INFO] org.apache.activemq:activemq-log4j-appender:jar:5.16.4-SNAPSHOT
Downloading from apache.snapshots: https://repository.apache.org/snapshots/org/apache/activemq/activemq-log4j-appender/5.16.4-SNAPSHOT/maven-metadata.xml
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.4-SNAPSHOT:compile
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |     +- log4j:log4j:jar:1.2.17:compile
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  +- org.apache.activemq:activemq-log4j-appender:jar:5.16.4-SNAPSHOT:test
[INFO] |  |  +- org.ops4j.pax.logging:pax-logging-log4j2:jar:1.11.9:test
[INFO] |  |     +- org.slf4j:slf4j-log4j12:jar:1.7.25:test
[INFO] |  |     +- log4j:log4j:jar:1.2.17:test
[INFO] |  +- org.slf4j:slf4j-log4j12:jar:1.7.25:compile
[INFO] |  +- log4j:log4j:jar:1.2.17:compile
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.4-SNAPSHOT:compile
[INFO] +- io.fabric8.insight:insight-log4j:jar:1.2.0.Beta4:compile
[INFO] ActiveMQ :: Log4j Appender ......................... SUCCESS [  0.004 s]
```

**Comment by gemmellr** [comment]
Yes that seems to be exactly the same as what I posted, just with a few more non-dependency related lines thrown in that your grep(/equivalent) also picked up.

**Comment by coheigea** [comment]
@gemmellr @jbonofre I could reproduce on Linux, but not on my Mac for some reason, weird! I updated the PR and now I don't see any log4j or slf4j-log4j on either Linux or Mac in the dependency tree.

**Comment by jbonofre** [comment]
Thanks ! I will take a look and test.

File: activemq-all/pom.xml [hunk 1]
```diff
@@ -14,7 +14,7 @@
   <parent>
     <groupId>org.apache.activemq</groupId>
     <artifactId>activemq-parent</artifactId>
-    <version>5.16.4-SNAPSHOT</version>
+    <version>5.16.5-SNAPSHOT</version>
```
Comment (gemmellr) [inline activemq-all/pom.xml:17]: PR needs rebased to remove the release commits deleted from the branch.

File: activemq-karaf/pom.xml [hunk 2]
```diff
@@ -42,12 +42,12 @@
     <dependency>
       <groupId>org.ops4j.pax.logging</groupId>
       <artifactId>pax-logging-api</artifactId>
-      <version>1.11.4</version>
+      <version>1.11.14</version>
```
Comment (gemmellr) [inline activemq-karaf/pom.xml:45]: Should probably add a property for this, perhaps dependencyManagement, to keep it consistent (I see from the dependency:trree there are a few different pax-logging versions in use in different modules).

**Review by gemmellr — COMMENTED** [review]
That does the trick, no sign of them in dependency:tree now, and double checked the built assembly doesnt contain the jar anymore but does have reload4j and slf4j-reload4j. Just a couple other minor comments.

```
$ mvn dependency:tree | grep log4j
[INFO] ------------< org.apache.activemq:activemq-log4j-appender >-------------
[INFO] --- maven-dependency-plugin:3.1.2:tree (default-cli) @ activemq-log4j-appender ---
[INFO] org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:compile
[INFO] |  +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:test
[INFO] |  |  +- org.ops4j.pax.logging:pax-logging-log4j2:jar:1.11.9:test
[INFO] +- org.apache.activemq:activemq-log4j-appender:jar:5.16.5-SNAPSHOT:compile
[INFO] +- io.fabric8.insight:insight-log4j:jar:1.2.0.Beta4:compile
```

**Comment by jbonofre** [comment]
@coheigea can you please rebase on updated `activemeq-5.16.x` branch ? It will trigger a new build (to test if LevelDB tests are OK).

File: activemq-all/pom.xml (hunk 1 above)
Comment (jbonofre) [reply activemq-all/pom.xml:17]: Agree, that's not up to date with the branch state.

File: activemq-karaf/pom.xml (hunk 2 above)
Comment (jbonofre) [reply activemq-karaf/pom.xml:45]: Agree, both activemq-karaf and karaf itest use pax-logging.

**Comment by jbonofre** [comment]
@coheigea while rebasing; can you also squash all commits in one. Else, I will do it myself.

**Comment by coheigea** [comment]
@jbonofre @gemmellr OK all done, thanks

**Comment by jbonofre** [comment]
@coheigea thanks; I'm doing a new pass and test.

**Comment by jbonofre** [comment]
Tests are failing, I'm checking if it's related to reload4j use.

**Comment by gemmellr** [comment]
There is a new 1.2.19 release today, probably worth using that also.

**Comment by gemmellr** [comment]
Plus looks like SLF4J 1.7.36 is being progressed for release just now.

**Comment by coheigea** [comment]
@gemmellr Thanks, I updated the PR to use 1.2.19.

**Comment by coheigea** [comment]
@jbonofre BTW all the itest fail for me, but also on activemq-5.16.x, so it's not related to this PR:

resolution:=mandatory [caused by: Unable to resolve org.apache.activemq.activemq-osgi/5.16.4.SNAPSHOT: missing requirement [org.apache.activemq.activemq-osgi/5.16.4.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=javax.annotation.meta)(version>=1.0.0)(!(version>=4.0.0)))"]]]]

**Comment by jbonofre** [comment]
@coheigea weird, it worked for me. Let me do new full test.

**Comment by coheigea** [comment]
@jbonofre I tried re-compiling activemq-5.16.x with JDK 11 but I see an error which seems related:

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.9.0:compile (default-compile) on project activemq-broker: Compilation failure: Compilation failure:
[ERROR] /Users/coheigea/src/apache/activemq/activemq-broker/src/main/java/org/apache/activemq/broker/util/LoggingBrokerPlugin.java:[21,24] package javax.annotation does not exist
[ERROR] /Users/coheigea/src/apache/activemq/activemq-broker/src/main/java/org/apache/activemq/broker/BrokerService.java:[51,24] package javax.annotation does not exist
[ERROR] /Users/coheigea/src/apache/activemq/activemq-broker/src/main/java/org/apache/activemq/broker/BrokerService.java:[52,24] package javax.annotation does not exist
[ERROR] /Users/coheigea/src/apache/activemq/activemq-broker/src/main/java/org/apache/activemq/broker/util/LoggingBrokerPlugin.java:[59,6] cannot find symbol
[ERROR]   symbol:   class PostConstruct
[ERROR]   location: class org.apache.activemq.broker.util.LoggingBrokerPlugin
[ERROR] /Users/coheigea/src/apache/activemq/activemq-broker/src/main/java/org/apache/activemq/broker/BrokerService.java:[582,6] cannot find symbol
[ERROR]   symbol:   class PostConstruct
[ERROR]   location: class org.apache.activemq.broker.BrokerService
[ERROR] /Users/coheigea/src/apache/activemq/activemq-broker/src/main/java/org/apache/activemq/broker/BrokerService.java:[796,6] cannot find symbol
[ERROR]   symbol:   class PreDestroy
[ERROR]   location: class org.apache.activemq.broker.BrokerService
[ERROR] -> [Help 1]

**Comment by jbonofre** [comment]
@coheigea i guess you didn't get my point : you have to use jdk8 max

**Comment by jbonofre** [comment]
Again jdk11 for runtime only, build is jdk8 max on 5.16.x. Jdk11 for build and run on 5.17.x

**Comment by coheigea** [comment]
Anyway, only 4 tests failed in the latest build, so the PR can be merged? All but TcpTransportInactiveDuringHandshakeTest pass for me locally.

**Comment by jbonofre** [comment]
Don't worry I'm on it. But I'm testing it on my build machines. I will merge soon.

**Comment by coheigea** [comment]
@jbonofre The itests fail for me with JDK8 on Linux as well by the way.

**Pull request merged**